File: GFG/Nine Divisors/Nine Divisors.java
class Solution{
    static long nineDivisors(long N){
        //Code Here
        long ans=0;
        int size=(int)Math.sqrt(N);
        int prime[]=new int[size+1];
        for(int i=1;i<=size;i++){
            prime[i]=i;
        }
        
        //fill sieve
        for(int i=2;i*i<=size;i++){
            if(prime[i]==i){
                for(int j=i*i;j<=size;j+=i){
                    if(prime[j]==j){
                        prime[j]=i;
                    }
                }
            }
        }
        //check for root n
        for(int i=2;i<=size;i++){
            int p=prime[i];
            
            int q=prime[i/p];
            
            if((p!= q && p*q==i && q!=1) ||
            (prime[i]==i && Math.pow(i,8)<=N)){
                ans+=1;
            }
        }
        return ans;
    }
}

File: GFG/Nine Divisors/Nine Divisors.cpp
class Solution{   
public:
    long long int nineDivisors(long long int N){
        //Code Here
        if(N < 36) return 0;
        
        long long size = sqrt(N);
        
        long long count = 0;
        
        vector<long long> sieve(size+1);
        
        for(int i = 1; i <= size; i++)
        {
            sieve[i] = i;
        }
        
        for(long long i = 2; i*i <= size; i++)
        {
            if(sieve[i] != i) continue;
            for(long long j = i*i; j <= size; j+=i)
            {
                if(sieve[j] == j)
                sieve[j] = i;
            }
        }
        
        for(long long i = 2; i <= size; i++)
        {
            long long p = sieve[i];
            
            long long q = sieve[i/p];
            
            if((p!=q && p*q == i && q !=1) || (sieve[i] == i && pow(i,8) <= N))
            {
                count++;
            }
        }
        return count;
    }
};
